godot-core/src/builtin/vector3.rs

```diff
@@ -4,75 +4,114 @@
+#[derive(Debug, Default, Clone, Copy, PartialEq)]
 #[repr(C)]
 pub struct Vector3 {
-    inner: Inner,
+    /// The vector's X component.
+    pub x: f32,
+    /// The vector's Y component.
+    pub y: f32,
+    /// The vector's Z component.
+    pub z: f32,
 }
 
+impl_vector_operators!(Vector3, f32, (x, y, z));
+impl_vector_index!(Vector3, f32, (x, y, z), Vector3Axis, (X, Y, Z));
+impl_common_vector_fns!(Vector3, f32);
+impl_float_vector_fns!(Vector3, f32);
+
 impl Vector3 {
-    pub fn new(x: f32, y: f32, z: f32) -> Self {
-        Self {
-            inner: Inner::new(x, y, z),
-        }
+    /// Returns a `Vector3` with the given components.
+    pub const fn new(x: f32, y: f32, z: f32) -> Self {
+        Self { x, y, z }
     }
-}
 
-impl GodotFfi for Vector3 {
-    ffi_methods! { type sys::GDExtensionTypePtr = *mut Self; .. }
-}
+    /// Returns a new `Vector3` with all components set to `v`.
+    pub const fn splat(v: f32) -> Self {
+        Self { x: v, y: v, z: v }
+    }
 
-impl std::fmt::Display for Vector3 {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        //let Inner {x, y, z} = self.inner;
-        //write!(f, "({x}, {y}, {z})")
-        self.inner.fmt(f)
+    /// Constructs a new `Vector3` from a [`Vector3i`].
+    pub const fn from_vector3i(v: Vector3i) -> Self {
+        Self { x: v.x as f32, y: v.y as f32, z: v.z as f32 }
     }
-}
 
-// ----------------------------------------------------------------------------------------------------------------------------------------------
+    /// Zero vector, a vector with all components set to `0.0`.
+    pub const ZERO: Self = Self::splat(0.0);
 
-type IInner = glam::IVec3;
+    /// One vector, a vector with all components set to `1.0`.
+    pub const ONE: Self = Self::splat(1.0);
 
-#[derive(Default, Copy, Clone, Debug, Eq, PartialEq)]
-#[repr(C)]
-pub struct Vector3i {
-    inner: IInner,
-}
+    /// Infinity vector, a vector with all components set to `INFIINTY`.
+    pub const INF: Self = Self::splat(f32::INFINITY);
+
+    /// Left unit vector. Represents the local direction of left, and the global direction of west.
+    pub const LEFT: Self = Self::new(-1.0, 0.0, 0.0);
 
-impl Vector3i {
-    pub fn new(x: i32, y: i32, z: i32) -> Self {
-        Self {
-            inner: IInner::new(x, y, z),
-        }
+    /// Right unit vector. Represents the local direction of right, and the global direction of east.
+    pub const RIGHT: Self = Self::new(1.0, 0.0, 0.0);
+
+    /// Up unit vector.
+    pub const UP: Self = Self::new(0.0, -1.0, 0.0);
+
+    /// Down unit vector.
+    pub const DOWN: Self = Self::new(0.0, 1.0, 0.0);
+
+    /// Forward unit vector. Represents the local direction of forward, and the global direction of north.
+    pub const FORWARD: Self = Self::new(0.0, 0.0, -1.0);
+
+    /// Back unit vector. Represents the local direction of back, and the global direction of south.
+    pub const BACK: Self = Self::new(0.0, 0.0, 1.0);
```

Comment:
"Global direction of north/east/south/west", how do you mean this? Is there such a terminology?

I think this depends on the game's coordinate/geography system, although I agree that XZ is usually the plane.

Reply:
Also copied from upstream, now removed.